tracer/src/Datadog.Trace/Span.cs

```diff
@@ -154,22 +155,58 @@ public override string ToString()
         /// <param name="key">The tag's key.</param>
         /// <param name="value">The tag's value.</param>
         /// <returns>This span to allow method chaining.</returns>
-        internal ISpan SetTag(string key, string value)
+        internal Span SetTag(string key, string value)
         {
             if (IsFinished)
             {
                 Log.Warning("SetTag should not be called after the span was closed");
                 return this;
             }
 
-            // some tags have special meaning
+            static void LogMissingTraceContext(string key, string value)
+            {
+                Log.Warning("Ignoring ISpan.SetTag({key}, {value}) because the span is not associated to a TraceContext.", key, value);
+            }
+
+            // since we don't expose a public API for setting trace-level attributes yet,
+            // allow setting them through any span in the trace.
+            // also, some "pseudo-tags" have special meaning, such as "manual.keep" and "_dd.measured".
             switch (key)
             {
+                case Trace.Tags.Env:
+                    if (Context.TraceContext == null)
+                    {
+                        LogMissingTraceContext(key, value);
+                        return this;
+                    }
```

Comment:
Are there _any_ cases in which we could _not_ have a TraceContext? If a span doesn't have a trace context, we'll throw on `Finish()` for example. Maybe this is a loophole we can close later?

Reply:
I don't think it ever happens for users, but we have [many tests](https://cs.github.com/DataDog/dd-trace-dotnet?q=%22new+Span%28%22) that create spans directly using `new Span(...)`, instead of using `Tracer.StartActive()` or `Tracer.StartSpan()`. We then try to add tags, read tags, serialize the span, etc, but none of those will have a `TraceContext`.

Many of changes I made to tests in this PR were specifically to create spans via a `Tracer` so they get an associated `TraceContext`, since that is now required to have the values for env, version, origin, sampling priority, etc.

I hope to rewrite the remaining tests someday and not allow null `TraceContext` anymore. Then we can remove all these null checks.